### ISSUE TITLE
Requesting WITH clause compatibility to JdbcPagingItemReader(PagingQueryProvider)

### ISSUE THREAD

**Issue by asalinasj** [spring-projects/spring-batch#3709]
I am currently working on Spring Batch performance when reading and writing to a DB2 datasource.  I have generally used Spring Batch with MySQL and compatibility and performance have been great when utilizing the JdbcPagingItemReader. When transitioning to DB2, the use of the WITH clause has become paramount when fetching data. When employing paging without the "WITH UR" statement, deadlocking can occur and performance drops. I'd like to request the addition of with clause compatibility in the PagingQueryProvider classes.

**Expected Behavior**
I expect the performance to increase and for the optimization to increase further with the use of local partitioning.

**Current Behavior**
Without the use of a WITH clause, performance drops heavily. Reading and writing 36,000 records takes around 45 seconds for the step to complete with the "WITH UR" clause and using prepared statements. The performance drops to 10-11 mins when not using the "WITH UR" clause.

<img width="446" alt="step" src="https://user-images.githubusercontent.com/25590457/81736464-9f641c00-9464-11ea-9d86-05463e68b923.png">

**Context**

**Comment by benas** [comment]
Yes, this seems to be a great performance improvement for DB2! We can consider introducing the change in one of the minor releases of v5 (we are currently focusing on other issues planned for 5.0). Pull requests are welcome!
